content/documentation/pages/6-recipes/7-scaling/2-autoscaling.md

```diff
@@ -0,0 +1,180 @@
+---
+path: 'recipes/scaling/autoscaling/'
+title: 'Autoscaling'
+description: 'Implement SCDF autoscaling with Prometheus AlertManager and SCDF Scale API'
+---
+
+# Autoscaling with Prometheus, Alertmanager and SCDF Scale API
+
+## Overview
+
+Autoscaling technique that extends the [Spring Cloud Data Flow Prometheus monitoring integration](http://localhost:8000/docs/2.4.0.SNAPSHOT/feature-guides/streams/monitoring/) to build an autoscaling controller for streaming data pipelines.
+It leverages the [Prometheus Alert Rules](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/) for defining `scale-out` and `scale-in` alerts based on pipeline metrics and actionates those alerts with [Prometheus AlertManager](https://prometheus.io/docs/alerting/alertmanager) and custom [webhook](https://github.com/prometheus/alertmanager), turning them into Data Flow Scale API calls.
+
+For a streaming data pipeline such as: `time | transform | log`, we will show how to use the delta of the streaming rates between the `time` and the `transform` applications to define scale alerts. When the so defined delta exceeds a certain threshold it fires a `scale-out` alert or a `scale-in` alert if the delta is none. This pseudocode illustrates the logic of such rules:
+
+```basic
+pipeline = time | transform | log
+delta-rate = rate(time) - rate(transform)
+IF delta-rate > 500 for 1 minute THEN fire ThroughputDeltaHigh
+IF delta-rate == 0 for 3 minutes THEN fire ZeroThroughputDelta
+```
+
+Following diagram shows the high level architecture.
+
+![SCDF autoscaling architecture](images/scdf-auto-scale-out-architecture.png)
+
+The alerting in Prometheus is separated into two parts: `Alert Rules` defined inside the Prometheus service and standalone `Alertmanager` service that receives and manages the alerts and in turn sends notification to external pre-registered webhooks.
+
+The [Alert Rules](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/) are based on the Prometheus Expression Language (PQL).
+They are used to define the scale alerts send to the `Alertmanager`.
+For example the scale-out alert rule definition looks something like this:
+
+```sql
+alert: ThroughputDeltaHigh
+expr:  avg(irate(spring_integration_send_seconds_count{application_name="time"}[1m])) by(stream_name) -
+       avg(irate(spring_integration_send_seconds_count{application_name="transform"}[1m])) by(stream_name) > 500
+for: 30s
+```
+
+<!--TIP-->
+
+Check [alert.rule.yml]() (and [here]() for the kubectl installation) to see the used alert rules definitions.
+
+<!--END_TIP-->
+
+The [Alertmanager](https://prometheus.io/docs/alerting/alertmanager) is a standalone service that manages the alerts, including silencing, inhibition, aggregation and sending out notifications to pre-configured webhooks.
+Through the [config.yml](), the Alertmanager configures webhooks such as the `AlertWebHookApplication`.
```

Comment:
>  [config.yml]() 

seems incomplete

Reply:
Indeed, because it depends on the https://github.com/spring-cloud/spring-cloud-dataflow-samples/pull/125 PR to be merged first, se we can reference the artifacts on github.